chain/chain/src/chain.rs

```diff
@@ -4165,6 +4269,12 @@ impl<'a> ChainUpdate<'a> {
     }
 }
 
+pub fn do_apply_chunks(
+    work: Vec<Box<dyn FnOnce() -> Result<ApplyChunkResult, Error> + Send>>,
+) -> Vec<Result<ApplyChunkResult, Error>> {
+    work.into_par_iter().map(|task| task()).collect::<Vec<_>>()
```

Comment:
:thinking: this `par_iter` in chain is inconsistent with a similar construct in runtime:

https://github.com/matklad/nearcore/blob/64e4ee198866135863975cb5b4daf0438e9adbbc/nearcore/src/runtime/mod.rs#L609-L616

Here, we use all the CPU cores to run the tasks.

There, we use half of the cores to run the tasks. 

If the two activities can actually happen concurrently, we will be using 1.5x threads total, which is clearly suboptimal. 

I wonder if we should change the runtime code to also just use global threadpool, but just limit the number of threads there at the startup.

Reply:
This is probably out of scope of this PR as this part is just refactoring. Can you create issue for this?

Reply:
https://github.com/near/nearcore/issues/4882

Reply:
@matklad I don't understand rayon very well, but is it correct that this code uses `work.len()` number of threads?

Reply:
I think it uses min(work.len(), number of cores) threads

Reply:
To be more precise:

Rayon has a global work-stealing thread pool. It's main primitive is scheduling a task for concurrent execution. The task then gets picked up by the most convenient thread from the thread pool. The threads dynamically redistribute the tasks among themselves.

So, what *this* code is doing is just slicing up the vector into a bunch of independent tasks. 

How exactly the tasks are mapped to threads depends on the details of the global threadpool. If we use default theradpool, and it is idle, then we'll see n_cpus threads churning through the tasks (and the threads won't be spawned just for this specific task, they will be pre-existing). In, in contrast, the thread pool is busy by executing tasks from some unrelated subsistem, this code might end up being executed by the single thread.